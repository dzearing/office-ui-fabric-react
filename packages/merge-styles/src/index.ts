export {
  IRawStyle,
  IStyle
} from './IStyle';

export {
  IStyleSet
} from './IStyleSet';

export {
  IFontFace,
  IFontWeight
} from './IRawStyleBase';

export {
  mergeStyles
} from './atomicMerge';

export {
  mergeStyleSets
} from './atomicMerge';

export {
  concatStyleSets
} from './concatStyleSets';

export {
  fontFace
} from './fontFace';

export {
  keyframes
} from './keyframes';

export {
  IStyleSheetConfig,
  InjectionMode,
  Stylesheet
} from './Stylesheet';
