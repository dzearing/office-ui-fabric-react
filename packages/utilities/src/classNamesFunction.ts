import { mergeStyleSets, IStyle, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
import { IStyleFunctionOrObject } from '@uifabric/merge-styles';

const MAX_CACHE_COUNT = 30;

// tslint:disable:no-any
const RetVal = '__retval__';
/**
 * Creates a getClassNames function which calls getStyles given the props, and injects them
 * into mergeStyleSets.
 *
 * Note that the props you pass in on every render should be in the same order and
 * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
 * these will cause extra recalcs to occur.
 */
export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
  getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
  styleProps?: TStyleProps
) => IProcessedStyleSet<TStyleSet> {
  let map: Map<any, any> = new Map<any, any>();

  const getClassNames = (
    styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
    styleProps: TStyleProps = {} as TStyleProps
  ): IProcessedStyleSet<TStyleSet> => {
    let current: Map<any, any> = map;
    let resultCount = 0;

    for (const propName in styleProps) {
      if (styleProps.hasOwnProperty(propName)) {
        let propValue: any = styleProps[propName];

        if (propValue === undefined) {
          propValue = '__undefined__';
        }
        if (propValue === null) {
          propValue = '__null__';
        }
        if (!current.has(propValue)) {
          current = current.set(propValue, new Map<any, any>());
        }
        current = current.get(propValue);
      }
    }

    if (!(current as any)[RetVal]) {
      if (styleFunctionOrObject === undefined) {
        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
      } else {
        const styleSet =
          styleFunctionOrObject &&
          (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);

        (current as any)[RetVal] = mergeStyleSets(styleSet as TStyleSet);
        resultCount++;
      }
    }

    if (resultCount > MAX_CACHE_COUNT) {
      if (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production') {
        throw new Error('Styles are being recalculated far too frequently.');
      }
      map.clear();
    }

    return (current as any)[RetVal] as IProcessedStyleSet<TStyleSet>;
  };

  return getClassNames;
}
