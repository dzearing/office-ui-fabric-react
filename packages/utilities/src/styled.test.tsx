import * as React from 'react';
import { styled } from './styled';
import * as renderer from 'react-test-renderer';
import { Customizer } from './Customizer';
import { IStyle, Stylesheet, InjectionMode, IStyleFunctionOrObject } from '@uifabric/merge-styles';
import { classNamesFunction } from './classNamesFunction';
import { memoizeFunction } from './memoize';

interface ITestStyles {
  root: IStyle;
}
interface ITestProps {
  cool?: boolean;
  styles?: IStyleFunctionOrObject<{}, ITestStyles>;
}

let _lastProps: ITestProps | undefined;
let _getStylesCount = 0;

const getClassNames = classNamesFunction<{}, ITestStyles>();

class TestBase extends React.Component<ITestProps> {
  public constructor(props: ITestProps) {
    super(props);
  }
  public render(): JSX.Element {
    _lastProps = this.props;
    const classNames = getClassNames(this.props.styles, this.props);

    return <div className={classNames.root} />;
  }
}

const getStyles = (props: ITestProps): ITestStyles => {
  _getStylesCount++;

  return {
    root: {
      background: props.cool ? 'blue' : 'red'
    }
  };
};

const getMemoizedStyles = memoizeFunction((cool?: boolean) => {
  _getStylesCount++;
  return {
    root: {
      background: cool ? 'blue' : 'red'
    }
  };
});

const Test = styled<ITestProps, {}, ITestStyles>(TestBase, getStyles, undefined, { scope: 'Test' });
const TestCached = styled<ITestProps, {}, ITestStyles>(TestBase, (props: ITestProps) => getMemoizedStyles(props.cool), undefined, {
  scope: 'Test'
});

describe('styled', () => {
  beforeEach(() => {
    _lastProps = undefined;
    _getStylesCount = 0;
    Stylesheet.getInstance().setConfig({
      injectionMode: InjectionMode.none
    });
    Stylesheet.getInstance().reset();
  });

  it('renders base styles (background red)', () => {
    const component = renderer.create(<Test />);

    // Test that defaults are the base styles (red).
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('allows user overrides (background green)', () => {
    const component = renderer.create(<Test styles={{ root: { background: 'green' } }} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('allows for contextual overrides (background yellow)', () => {
    const component = renderer.create(
      <Customizer
        scopedSettings={{
          Test: {
            styles: {
              root: {
                background: 'yellow'
              }
            }
          }
        }}
      >
        <Test />
      </Customizer>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('allows for contextual overrides mixed under user overrides (background yellow, color red)', () => {
    const component = renderer.create(
      <Customizer
        scopedSettings={{
          Test: {
            styles: {
              root: {
                background: 'yellow'
              }
            }
          }
        }}
      >
        <Test styles={{ root: { color: 'red' } }} />
      </Customizer>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('can process style props (background blue)', () => {
    const component = renderer.create(<Test cool />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('can produce different styles object if not cached', () => {
    const component = renderer.create(<Test cool />);
    const originStyles = _lastProps!.styles;

    component.update(<Test cool />);
    expect(_lastProps!.styles).not.toBe(originStyles);
  });

  it('can produce same styles object if cached', () => {
    const component = renderer.create(<TestCached cool />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_getStylesCount).toEqual(1);

    // Calling again does not re-evaluate styles.
    component.update(<TestCached cool />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_getStylesCount).toEqual(1);

    // Not passing in "cool" should cause re-evaluation of styles.
    component.update(<TestCached />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_getStylesCount).toEqual(2);

    // Passing the same props in again should skip re-evaluation.
    component.update(<TestCached />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_getStylesCount).toEqual(2);

    // Passing in custom styling should not re-evaluate memoized styles.
    const styles = { root: { color: 'red' } };
    component.update(<TestCached cool styles={styles} />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_getStylesCount).toEqual(2);
  });
});
