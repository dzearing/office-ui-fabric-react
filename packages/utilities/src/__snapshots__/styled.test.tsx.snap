// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled allows for contextual overrides (background yellow) 1`] = `
<div
  className=

      {
        background: yellow;
      }
/>
`;

exports[`styled allows for contextual overrides mixed under user overrides (background yellow, color red) 1`] = `
<div
  className=

      {
        background: yellow;
        color: red;
      }
/>
`;

exports[`styled allows user overrides (background green) 1`] = `
<div
  className=

      {
        background: green;
      }
/>
`;

exports[`styled can process style props (background blue) 1`] = `
<div
  className=

      {
        background: blue;
      }
/>
`;

exports[`styled can produce same styles object if cached 1`] = `
<div
  className=

      {
        background: blue;
      }
/>
`;

exports[`styled can produce same styles object if cached 2`] = `
<div
  className=

      {
        background: blue;
      }
/>
`;

exports[`styled can produce same styles object if cached 3`] = `
<div
  className=

      {
        background: red;
      }
/>
`;

exports[`styled can produce same styles object if cached 4`] = `
<div
  className=

      {
        background: red;
      }
/>
`;

exports[`styled can produce same styles object if cached 5`] = `
<div
  className=

      {
        background: blue;
        color: red;
      }
/>
`;

exports[`styled renders base styles (background red) 1`] = `
<div
  className=

      {
        background: red;
      }
/>
`;
